crates/holochain/src/conductor/conductor.rs

```diff
@@ -2415,6 +2457,83 @@ mod service_impls {
+        /// Revoke an agent's key pair for all cells of an app. The agent key becomes invalid and
+        /// source chains of all cells of the app are made read-only.
+        pub async fn revoke_agent_key_for_app(
+            self: Arc<Self>,
+            agent_key: AgentPubKey,
+            app_id: InstalledAppId,
+        ) -> ConductorResult<(ActionHash, KeyRegistration)> {
+            let dpki_service = self
+                .running_services()
+                .dpki
+                .ok_or(ConductorError::DpkiError(
+                    DpkiServiceError::DpkiNotInstalled,
+                ))?;
+            let dpki_state = dpki_service.state().await;
+            let timestamp = Timestamp::now();
+            let key_state = dpki_state
+                .key_state(agent_key.clone(), timestamp.clone())
+                .await?;
+            match key_state {
+                KeyState::NotFound => Err(ConductorError::DpkiError(
+                    DpkiServiceError::DpkiAgentMissing(agent_key),
+                )),
+                KeyState::Invalid(_) => Err(ConductorError::DpkiError(
+                    DpkiServiceError::DpkiAgentInvalid(agent_key, timestamp),
+                )),
+                KeyState::Valid(_) => {
+                    // get action hash of key registration
+                    let key_meta = dpki_state.query_key_meta(agent_key.clone()).await?;
+                    // sign revocation request
+                    let signature = dpki_service
+                        .cell_id
+                        .agent_pubkey()
+                        .sign_raw(
+                            &self.keystore,
+                            key_meta.key_registration_addr.get_raw_39().into(),
+                        )
+                        .await
+                        .map_err(|e| DpkiServiceError::Lair(e.into()))?;
+                    let signature = deepkey_roundtrip_backward!(Signature, &signature);
+
+                    // disable app, revoke key, close all cells' chains and enable app again
+                    self.clone()
+                        .disable_app(app_id.clone(), DisabledAppReason::DeleteAgentKey)
+                        .await?;
+                    let revocation = dpki_state
+                        .revoke_key(RevokeKeyInput {
+                            key_revocation: KeyRevocation {
+                                prior_key_registration: key_meta.key_registration_addr,
+                                revocation_authorization: vec![(0, signature)],
+                            },
+                        })
+                        .await?;
+
+                    // close chain of all cells of this app
+                    let state = self.get_state().await?;
+                    let app = state.get_app(&app_id)?;
+                    let close_chain_requests = app
+                        .all_cells()
+                        .map(|cell_id| Self::close_chain(self.clone(), cell_id));
```

Comment:
Regardless of if you close chain or not, you also need to delete the key on the app chain.

Reply:
I see what you mean, but I think the issue needs to be clearer in that regard, or some spec should detail this.